pairing/bn256/point.go

```diff
@@ -72,11 +99,61 @@ func (p *pointG1) Embed(data []byte, rand cipher.Stream) kyber.Point {
 	//   filled with random values, i.e., x = rand || data || len(data).
 	// - Use the Tonelli-Shanks algorithm to compute the y-coordinate.
 	// - Convert the new point to Jacobian coordinates and set it as p.
-	panic("bn256.G1: unsupported operation")
+
+	// How many bytes to embed?
+	dl := p.EmbedLen()
+	if dl > len(data) {
+		dl = len(data)
+	}
+
+	intCurveB := gfpToBig(curveB)
+
+	for {
+		// Pick a random point, with optional embedded data
+		var b [32]byte
+		rand.XORKeyStream(b[:], b[:])
+		if data != nil {
+			b[0] = byte(dl)       // Encode length in low 8 bits
+			copy(b[1:1+dl], data) // Copy in data to embed
+		}
+		x := new(big.Int).SetBytes(b[:])
+
+		xxx := new(big.Int).Mul(x, x)
+		xxx.Mul(xxx, x)
+		xxx.Mod(xxx, Order)
+
+		t := new(big.Int).Add(xxx, intCurveB)
+		y := new(big.Int).ModSqrt(t, Order)
+		if y != nil {
+			p.g.x = *bigToGfp(x)
+			p.g.y = *bigToGfp(y)
+			p.g.z = *newGFp(1)
+			q := newPointG1()
+			t := q.Sub(p, p)
+			if t.Equal(q.Null()) {
```

Comment:
You can actually use the already implemented function `p.g.IsOnCurve()`

Reply:
I added this and realized Order was the incorrect number to modulo operations by so its in mod P now. I also had to change the constant p to P because the receivers on those points is p.